File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        int sum = 0;
        ListNode* dummy = new ListNode();
        ListNode* curr = dummy;
        while(l1 or l2 or carry){
            int a = l1 ? l1->val: 0;
            int b = l2 ? l2->val:0;
            sum = a+b+carry;
            carry = sum/10;
            sum = sum%10;
            ListNode* node = new ListNode(sum);
            curr->next = node;
            curr = curr->next;
            if(l1)
                l1=l1->next;
            if(l2)
                l2= l2->next;
        }
        return dummy->next;
    }
};

File: 0002-add-two-numbers/0002-add-two-numbers.py
# TC- O(max(m,n))
# SC- O(1)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        curr = dummy
        carry = 0
        while(l1 or l2 or carry):
            a = l1.val if l1 else 0
            b = l2.val if l2 else 0
            sum = a + b + carry
            carry = sum//10
            sum = sum%10
            curr.next = ListNode(sum)
            curr = curr.next
            if l1:
                l1 = l1.next 
            if l2:
                l2 = l2.next

        return dummy.next
        